### PR TITLE
Feature/with session for routes

### LINKED ISSUE #371: Add dhis2 session validation to routes

### PR DESCRIPTION
**Pull request by xurxodev**
### :pushpin: References

* **Issue:** Closes #371

### :memo: Implementation

- I have renamed Authorization component to WithAutorization
- I have created WithSession component, this component to realize a request to /me endpoint, if the response is ok then return children else return a login link text as index.js
- I have created RouteWithSession and RouteWithSessionAndAuth to simplify root composition.

### :art: Screenshots
Captura de pantalla 2020-02-12 a las 12 20 27

### :fire: Is there anything the reviewer should know to test it?

* **use case 1** *: When a dhis2 session is finished to navigate a new route should show the login message.
* **use case 2** *: Authorization logic should work as before, showing unauthorize message if user has no permissions by share settings or to view delete objects typing the url.

### :bookmark_tabs: Others

- Any change in the [GUI library](https://github.com/EyeSeeTea/d2-ui-components)? If so, what branch/PR?

### CODE REVIEW

File: src/components/auth/RouteWithSession.tsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import { Route, RouteComponentProps } from "react-router-dom";
+import WithSession from "./WithSession";
+
+export interface RouteWithSessionProps {
+    render: (props: RouteComponentProps<any>) => React.ReactNode;
```
Comment (tokland) [inline src/components/auth/RouteWithSession.tsx]: All RouteComponentProps arguments have defaults, so we can simply write `RouteComponentProps`

File: src/components/auth/RouteWithSession.tsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    path: string | string[];
```
Comment (tokland) [inline src/components/auth/RouteWithSession.tsx]: Path is optional for the default path, so `path?: ...`

File: src/components/auth/RouteWithSessionAndAuth.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import React from "react";
+import { Route, RouteComponentProps } from "react-router-dom";
+import WithSession from "./WithSession";
+import WithAuthorization from "./WithAuthorization";
+import { RouteWithSessionProps } from "./RouteWithSession";
+
+export interface RouteWithSessionAndAuthProps extends RouteWithSessionProps {
+    authorize: (props: RouteComponentProps<any>) => Promise<boolean>;
```
Comment (tokland) [inline src/components/auth/RouteWithSessionAndAuth.tsx]: Same than before, we can remove any

File: src/components/auth/RouteWithSession.tsx (hunk 1 above)
Comment (xurxodev) [reply src/components/auth/RouteWithSession.tsx]: done!

File: src/components/auth/RouteWithSession.tsx (hunk 2 above)
Comment (xurxodev) [reply src/components/auth/RouteWithSession.tsx]: done!

File: src/components/auth/RouteWithSessionAndAuth.tsx (hunk 3 above)
Comment (xurxodev) [reply src/components/auth/RouteWithSessionAndAuth.tsx]: done!

**Comment by tokland** [comment]
Replaced d2 by d2-api so we have the component fully typed.

**Pull request merged**